Add Snappy Compression

Bhargavasomu opened the pull request:
### What was wrong?
Fixes Issue #57


### How was it fixed?
* As of now, the Snappy Compression is made default for the transfer of messages in the Clients. In other words at this stage, this would work only if the following assumption holds, 
`All the clients in the Ethereum Network support Snappy Compression`.
* The Next step of this would be to check if the `other Node` which we are communicating with, supports `Snappy Compression` or not. And we should get this information in the `Handshake` itself. Only if it is supported, then we should send `Snappy Compressed Messages`. Else we shouldn't `Compress` the messages.
* Also `try/catch` exceptions should be added wherever we are using 
`Snappy compression/decompression`.

### References
* Mainly Referred [this](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-706.md) `EIP`
* Heavily Inspired by `geth` mainly from the functionalities `ReadMsg` and `WriteMsg` of 
`go-ethereum/p2p/rlpx.go`


#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://www.rd.com/wp-content/uploads/2017/10/203_The-Dogist-Puppies-760x506.jpg)

Bhargavasomu commented:
Sorry for the spam, but I have a fix for this. I modified the implementation so that `Snappy Compression/Decompression` is never used for the `Hello Command` (Which is used during `Handshake`). This provides an extra layer of security because as per the specs also, `Snappy Compression` should not be a part of the `Handshake`. And it also solved the above stated issue.

pipermerriam commented on p2p/peer.py:
```diff
@@ -496,6 +497,12 @@ def process_msg(self, cmd: protocol.Command, msg: protocol.PayloadType) -> None:
+        # Check whether to support Snappy Compression or not
+        # based on other peer's p2p protocol version
+        if msg['version'] >= SNAPPY_PROTOCOL_VERSION:
+            protocol.Command.snappy_support = True
```
This approach isn't going to work as it mutates the global `Command` class.

Instead what I believe we need to do is pass this information into the `Protocol` class itself here:

https://github.com/ethereum/trinity/blob/c4e387465b222f8a870966f0eacf7a68fa5aee7a/p2p/protocol.py#L154-L157

This will allow the value to be passed down into the `Command` class instances that are stored on the `Protocol` class itself.

Bhargavasomu replied:
@pipermerriam I actually *intended* it to mutate the global `Command` class. The reasoning that I came up for this action is as follows.

Once the Handshake is complete, the peer comes to a decision of whether to support or not support `snappy compression`. From that moment onwards, every `command` (Irrespective of the `protocol`) that the peer sends to some other remote has to be `compressed` if the peers decide to support `Snappy Compression`. Hence that's why I wanted to change the class globally.

pipermerriam commented:
@Bhargavasomu take a look at db9efb7

It updates your approach to remove the mutation.  It's mostly untested so I'm hoping you can take a look and clean up the failures.

Bhargavasomu commented on p2p/protocol.py:
```diff
@@ -48,13 +48,12 @@ class Command:
     _logger: logging.Logger = None
 
-    def __init__(self, cmd_id_offset: int) -> None:
+    def __init__(self, cmd_id_offset: int, snappy_support: bool=False) -> None:
```
Though we have set the default value of `snappy_support` to `False`, we should `raise` an `Exception` if the 
`Hello Command` is instantiated with `snappy_support` as `True`, because `snappy compression` shouldn't be enabled at the Handshake level. This would just be an extra layer of security.

pipermerriam commented:
@Bhargavasomu I believe that the only way to do multiple connections is to use a `PeerPool`.  The *concept* of being connected to multiple peers is only valid at the `PeerPool` level as our `Peer` classes only represent a 1:1 connection between two peers.

If you're able to test this cleanly :+1: but I'm also ok without that specific functionality being tested.

Bhargavasomu commented:
@pipermerriam I have added the very basic tests in the context of `Snappy Compression`. Please let me know if I need to change anything more. Thankyou.

carver commented on .circleci/config.yml:
```diff
@@ -46,6 +49,9 @@ geth_steps: &geth_steps
           - cache-v1-{{ arch }}-{{ .Environment.CIRCLE_JOB }}-{{ checksum "setup.py" }}-{{ checksum "tox.ini" }}
+    - run:
+        name: install libsnappy-dev
+        command: sudo apt install -y libsnappy-dev
```
This line belongs in the quickstart, too: https://trinity-client.readthedocs.io/en/latest/guides/trinity/quickstart.html#installing-on-ubuntu
(source in docs/guides/quickstart.rst)

If you happen to know how to do it on the mac, that section update would be cool too.

carver commented on p2p/protocol.py:
```diff
@@ -48,9 +50,10 @@ class Command:
     _logger: logging.Logger = None
 
-    def __init__(self, cmd_id_offset: int) -> None:
+    def __init__(self, cmd_id_offset: int, snappy_support: bool=False) -> None:
```
Maybe we could even make this a required field instead of defaulting to `False`. It could help us catch any initializations we might have missed.

carver commented on p2p/protocol.py:
```diff
@@ -152,12 +175,13 @@ class Protocol:
     _logger: logging.Logger = None
 
-    def __init__(self, peer: 'BasePeer', cmd_id_offset: int) -> None:
+    def __init__(self, peer: 'BasePeer', cmd_id_offset: int, snappy_support: bool=False) -> None:
```
Same here, maybe make this field required.

Bhargavasomu commented:
@carver @pipermerriam I have addressed all the above issues.
* Now `snappy_support` variable has been made `compulsory` instead of `default`.
* I have also updated the docs. Since I am not a `Mac` user, I haven't tested the commands stated in the `Mac` section of the docs, so if either of you are Mac users, could you please test it on your local machines before merging.

Thankyou.

pipermerriam commented:
cc @njgheorghita or @kclowes as I know both of you develop on OSX.  One of you mind running through the docs section in this PR and making sure everything seems to work as expected?  Being able to run the section of the tests in this PR that deals with snappy is probably sufficient/adequate.

pipermerriam commented:
@carver leaving to you for final merge decision.  Looks :+1: to me.
